File: webapp/backend/app/telemetry/sources/external_gpx.py
"""Telemetry source for cameras with no embedded GPS track.

Many action cams (DJI, Insta360, older GoPros) and most external GPS
trackers (phone apps, Garmin, Polar) don't embed GPMD in the video at all —
the user has a separate .gpx file instead. This source aligns that GPX
track onto the video's own timeline using either an explicit offset or the
video's real-world start time (matched against GPX point timestamps).
"""

from __future__ import annotations

from datetime import datetime
from pathlib import Path

import gpxpy
import numpy as np
import pandas as pd

from app.laps.detection import haversine_distance_m
from app.telemetry.resample import resample_to_grid
from app.telemetry.sources.base import TelemetryResult, empty_result


def load_gpx_points(gpx_path: Path) -> pd.DataFrame:
    """Flattens every track/segment point in a GPX file into one DataFrame."""
    with open(gpx_path, "r", encoding="utf-8") as f:
        gpx = gpxpy.parse(f)

    rows = []
    for track in gpx.tracks:
        for segment in track.segments:
            for point in segment.points:
                rows.append({
                    "timestamp": point.time,
                    "lat": point.latitude,
                    "lon": point.longitude,
                    "ele": point.elevation,
                })

    df = pd.DataFrame(rows)
    if df.empty or df["timestamp"].isna().any():
        return df
    return df.sort_values("timestamp").reset_index(drop=True)


def compute_speed_kmh(df: pd.DataFrame) -> pd.Series:
    """Derives speed from consecutive GPX points (most consumer GPX files
    don't carry a <speed> extension, so this is computed, not read)."""
    lat = df["lat"].to_numpy()
    lon = df["lon"].to_numpy()
    seconds = df["timestamp"].apply(lambda t: t.timestamp()).to_numpy()

    dist_m = np.zeros(len(df))
    dist_m[1:] = haversine_distance_m(lat[:-1], lon[:-1], lat[1:], lon[1:])
    dt = np.diff(seconds, prepend=seconds[0] if len(seconds) else 0)
    dt[dt <= 0] = np.nan

    speed_ms = dist_m / dt
    speed_ms = np.nan_to_num(speed_ms, nan=0.0)
    return pd.Series(speed_ms * 3.6, index=df.index)


class ExternalGpxSource:
    """TelemetrySource backed by a standalone GPX file, time-aligned to the video."""

    def __init__(self, gpx_path: Path, target_fps: float = 30.0, offset_sec: float = 0.0, video_start_time: datetime | None = None):
        self.gpx_path = Path(gpx_path)
        self.target_fps = target_fps
        self.offset_sec = offset_sec
        self.video_start_time = video_start_time

    def extract(self, video_path: Path, duration_sec: float) -> TelemetryResult:
        points = load_gpx_points(self.gpx_path)
        if points.empty:
            return empty_result("external_gpx")

        points["speed"] = compute_speed_kmh(points)

        if self.video_start_time is not None:
            points["time"] = (points["timestamp"] - self.video_start_time).dt.total_seconds()
        else:
            first_ts = points["timestamp"].iloc[0]
            points["time"] = (points["timestamp"] - first_ts).dt.total_seconds() + self.offset_sec

        windowed = points[(points["time"] >= 0) & (points["time"] <= duration_sec)]
        if len(windowed) < 2:
            return empty_result("external_gpx")

        resampled = resample_to_grid(windowed, duration_sec, self.target_fps, ["lat", "lon", "speed"])
        resampled["lat_g"] = 0.0
        resampled["lon_g"] = 0.0
        return TelemetryResult(df=resampled, source_name="external_gpx", has_accel=False)

File: frontend/src/lib/telemetry/externalGpx.ts
/**
 * Telemetry source for cameras with no embedded GPS track (a standalone GPX
 * file instead). Ported from backend/app/telemetry/sources/external_gpx.py
 * -- see that file's docstring for the alignment rationale.
 *
 * `gpxpy`'s full XML parsing is replaced with a targeted regex scan over
 * `<trkpt>` blocks rather than `DOMParser` -- a real-world GPX track can run
 * to tens of thousands of points, and a full XML DOM parse (jsdom measured
 * well over a minute on this module's own 8.7k-point test fixture) doesn't
 * scale the way this simple, regular, attribute-plus-two-child-elements
 * structure needs to. GPX itself doesn't require anything a DOM would buy
 * here: no nesting below `<trkpt>`, no namespaced siblings to disambiguate.
 */
import { haversineDistanceM } from '../laps/detection'
import { resampleToGrid, smoothSpeedOutliers } from './resample'
import { emptyResult, type TelemetryResult } from './base'

export interface GpxPoint {
  timestamp: Date
  lat: number
  lon: number
  ele: number | null
}

const TRKPT_RE = /<trkpt\b([^>]*)>([\s\S]*?)<\/trkpt>/g
const LAT_ATTR_RE = /\blat="([^"]+)"/
const LON_ATTR_RE = /\blon="([^"]+)"/
const TIME_RE = /<time>([^<]+)<\/time>/
const ELE_RE = /<ele>([^<]+)<\/ele>/

/** Flattens every track/segment point in a GPX file into one array, sorted
 * by timestamp. */
export function loadGpxPoints(gpxText: string): GpxPoint[] {
  const points: GpxPoint[] = []
  for (const match of gpxText.matchAll(TRKPT_RE)) {
    const [, attrs, body] = match
    const latMatch = LAT_ATTR_RE.exec(attrs)
    const lonMatch = LON_ATTR_RE.exec(attrs)
    const timeMatch = TIME_RE.exec(body)
    if (!latMatch || !lonMatch || !timeMatch) continue
    const timestamp = new Date(timeMatch[1])
    if (Number.isNaN(timestamp.getTime())) continue
    const eleMatch = ELE_RE.exec(body)
    points.push({
      timestamp,
      lat: Number(latMatch[1]),
      lon: Number(lonMatch[1]),
      ele: eleMatch ? Number(eleMatch[1]) : null,
    })
  }

  points.sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime())
  return points
}

/** Derives speed from consecutive GPX points (most consumer GPX files don't
 * carry a <speed> extension, so this is computed, not read). Position-delta
 * speed is much noisier than a GPS chip's own reported speed, so it's
 * smoothed the same way as the other sources. */
export function computeSpeedKmh(points: GpxPoint[]): number[] {
  const lat = points.map((p) => p.lat)
  const lon = points.map((p) => p.lon)
  const seconds = points.map((p) => p.timestamp.getTime() / 1000)

  const distM = new Array(points.length).fill(0)
  for (let i = 1; i < points.length; i++) {
    distM[i] = haversineDistanceM(lat[i - 1], lon[i - 1], lat[i], lon[i]) as number
  }

  const dt = seconds.map((s, i) => (i === 0 ? s - (seconds[0] ?? 0) : s - seconds[i - 1]))
  const speedMs = distM.map((d, i) => {
    const denom = dt[i]
    return denom > 0 ? d / denom : 0.0
  })
  const speedKmh = speedMs.map((v) => v * 3.6)
  return smoothSpeedOutliers(speedKmh, 5, 5.0, seconds)
}

export interface ExternalGpxOptions {
  targetFps?: number
  offsetSec?: number
  videoStartTime?: Date
}

export function extractExternalGpx(gpxText: string, durationSec: number, opts: ExternalGpxOptions = {}): TelemetryResult {
  const targetFps = opts.targetFps ?? 30.0
  const offsetSec = opts.offsetSec ?? 0.0

  const points = loadGpxPoints(gpxText)
  if (points.length === 0) return emptyResult('external_gpx')

  const speed = computeSpeedKmh(points)
  const referenceTime = opts.videoStartTime ?? points[0].timestamp

  const withTime = points.map((p, i) => ({
    lat: p.lat,
    lon: p.lon,
    speed: speed[i],
    time: (p.timestamp.getTime() - referenceTime.getTime()) / 1000 + offsetSec,
  }))

  const windowed = withTime.filter((p) => p.time >= 0 && p.time <= durationSec)
  if (windowed.length < 2) return emptyResult('external_gpx')

  const resampled = resampleToGrid(windowed, durationSec, targetFps, ['lat', 'lon', 'speed'])
  const rows = resampled.map((r) => ({
    time: r.time,
    lat: r.lat,
    lon: r.lon,
    speed: r.speed,
    lat_g: 0.0,
    lon_g: 0.0,
  }))
  return { rows, sourceName: 'external_gpx', hasAccel: false }
}
